qiskit/primitives/base_estimator.py

```diff
@@ -0,0 +1,199 @@
+        psi2_H2_result = e(theta2, shots=1024, grouping=[(1, 1)])
+        print(psi2_H2_result)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi1(theta3)|H1|psi1(theta3)> ]
+        psi1_H1_result2 = e([theta1, theta3], shots=1024, grouping=[(0, 0)])
+        print(psi1_H1_result2)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi2(theta2)|H2|psi2(theta2)>, <psi1(theta3)|H3|psi1(theta3)> ]
+        psi12_H23_result = e([theta1, theta2, theta3], shots=1024, grouping=[(0, 0), (1, 1), (0, 2)])
+        print(psi12_H23_result)
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Optional, Union
+
+from qiskit.circuit import QuantumCircuit, Parameter
+from qiskit.circuit.parametertable import ParameterView
+from qiskit.quantum_info import SparsePauliOp
+
+from .estimator_result import EstimatorResult
+
+
+@dataclass(frozen=True)
+class Group:
+    """The dataclass represents indices of circuit and observable."""
+
+    circuit_index: int
+    observable_index: int
+
+
+class BaseEstimator(ABC):
+    """Estimator base class.
+
+    Base class for Estimator that estimates expectation values of quantum circuits and observables.
+    """
+
+    def __init__(
+        self,
+        circuits: list[QuantumCircuit],
+        observables: list[SparsePauliOp],
+        parameters: Union[ParameterView, list[Parameter]],
+    ):
+        """
+        Args:
+            circuits (list[QuantumCircuit]): quantum circuits that represents quantum states
+            observables (list[SparsePauliOp]): observables
+            parameters (Union[ParameterView, list[Parameter]]): parameters of quantum circuits
+        """
+        self._circuits = circuits
+        self._observables = observables
+        self._parameters = parameters
+
+    @abstractmethod
+    def __enter__(self):
+        ...
+
+    @abstractmethod
+    def __exit__(self, ex_type, ex_value, trace):
+        ...
+
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        cls.__call__ = cls.run
+
+    @property
+    def circuits(self) -> list[QuantumCircuit]:
+        """Quantum circuits that represents quantum states.
+
+        Returns:
+            list[QuantumCircuit]: quantum circuits
+        """
+        return self._circuits
+
+    @property
+    def observables(self) -> list[SparsePauliOp]:
+        """Observables
+
+        Returns:
+            list[SparsePauliOp]: observables
+        """
+        return self._observables
+
+    @property
+    def parameters(self) -> Union[ParameterView, list[Parameter]]:
+        """Parameters of quantum circuits
+
+        Returns:
+            Union[ParameterView, list[Parameter]]: Parameter list of the quantum circuits
+        """
+        return self._parameters
+
+    @abstractmethod
+    def run(
+        self,
+        parameters: Optional[Union[list[float], list[list[float]]]] = None,
```

Comment:
```suggestion
        parameters: list[list[float]],
```
I don't see how it can be optional, and I prefer to always take it nested, even if just one set of params.

Reply:
Optional presumably when you send a concrete non-parameterized circuit.

But I think for just one parameter vector it is less tedious for the user to write it as a single list? You don't like the type-checking? We have this kind of patter in different parts of qiskit, for example transpile() takes either a single circuit or a list of circuits.

Reply:
On the subject of allowing both `List` and `List[List]`: Allowing both structures in this can lead to tricky inference issues. If I were to pass `[]`, should that be interpreted as an empty list of type `List[List[float]]` where the inner list has an unknown number of elements, or of type `List[float]` with 0 elements?  It can validly be promoted to `[[]]` or to `[]`, and you can't tell which, and the semantics are different: one implies that it acts once, and another implies it acts zero times.

There may be ways around that (perhaps neither is valid?) for the use-cases we immediately have in mind, but it's a potential minefield if the interface were to expand.

On a side note: if we're planning to make type hints a part of the interface, it's worth thinking about whether the inputs should be `List` - is there a good reason to reject `tuple`, `ndarray`, or any other sequence?  Python's `List` is an invariant type because it's mutable, so if I wrote:
```python
def f(x: list[float]):
    pass

my_list = [1, 2, 3]
f(my_list)
```
it'd be a type error, because "list of int" is incompatible with "list of float" (because e.g. `List[int].append(1.5)` is invalid).  If the interface doesn't need to mutate its input (and it really shouldn't), it probably wants to use a covariant type instead, like `Sequence`:
```python
from typing import Sequence

def f(x: Sequence[float]):
    pass

my_list = [1, 2, 3]
f(my_list)
```
is fine.

Reply:
> Optional presumably when you send a concrete non-parameterized circuit.

In that case it can be a zero-length list

> But I think for just one parameter vector it is less tedious for the user to write it as a single list? You don't like the type-checking? We have this kind of patter in different parts of qiskit, for example transpile() takes either a single circuit or a list of circuits.

Yeah, but like Jake said it makes the type checking tricky and to test it properly you have to write test cases for both versions, and the shape of the return type changes depending on the input, etc. I agree we do it elsewhere but that was largely where we started with something that was a single element and changed it to a list as an afterthought. We have a chance to get this right from the beginning